src/Core/System/SalesChannel/Context/BaseSalesChannelContextFactory.php

```diff
@@ -279,19 +278,30 @@ private function getCashRounding(CurrencyEntity $currency, ShippingLocation $shi
         return [$currency->getItemRounding(), $currency->getTotalRounding()];
     }
 
-    private function getLanguageInfo(?LanguageCollection $languages, string $currentLanguageId): LanguageInfo
+    private function getLanguageInfo(string $currentLanguageId, Context $context): LanguageInfo
     {
-        $currentLanguage = $languages?->get($currentLanguageId);
-        if ($currentLanguage === null) {
+        $criteria = new Criteria([$currentLanguageId]);
+        $criteria->addFields([
+            'id',
+            'name',
+            'translationCode.code',
+            'translationCode.name',
+            'locale.code',
+            'locale.name',
+        ]);
+
+        $currentLanguage = $this->languageRepository->search($criteria, $context)->getEntities()->get($currentLanguageId);
+        if (!$currentLanguage instanceof PartialEntity) {
             throw SalesChannelException::languageNotFound($currentLanguageId);
         }
 
-        $locale = $currentLanguage->getTranslationCode() ?? $currentLanguage->getLocale();
-        \assert($locale !== null, 'At least the localeId is required, so the fallback should never be null');
+        $locale = $currentLanguage->get('translationCode') ?? $currentLanguage->get('locale');
+        \assert($locale instanceof PartialEntity, 'At least the localeId is required, so the fallback should never be null');
 
         return new LanguageInfo(
-            $currentLanguage->getTranslation('name') ?? $currentLanguage->getName(),
-            $locale->getCode(),
+            $currentLanguage->get('name'),
```

Comment:
I would move all language information to this object and provide a LanguageInfoLoader.

I would miss:
- language.id
- language-id-chain
- locale.id

And in the SalesChannelContext / Context, i would delegate the language data calls to the language info object

```
class Context
{
    public function getLanguageId() {
        return this->languageInfo->languageId;
    }
 
    public function getLanguageIdChain() {
       return this->languageInfo->languageChain
   }
}
...
```

Reply:
already did that https://github.com/shopware/shopware/blob/trunk/src/Core/System/SalesChannel/Context/ContextFactory.php
but missed to use it in other places. 
I will revisit that, thanks for your comments!

Reply:
I decided to separate that into another PR